gemrb/core/Scriptable/Scriptable.cpp

```diff
@@ -2143,49 +2145,91 @@ unsigned char Movable::GetNextFace()
+	int stepLength = 566;
+	if (core->HasFeature(GF_BREAKABLE_WEAPONS)) { // BG1
+		stepLength = 425;
 	}
+	const double maxRadius = 2.0;
+	if (!path) return true;
 	if (!time) time = core->GetGame()->Ticks;
-	if (!walk_speed) {
+	if (!step) {
+		step = path;
+		timeStartStep = time;
+	}
+
+	if (!walkAnimFrames) {
 		// zero speed: no movement
 		timeStartStep = time;
 		StanceID = IE_ANI_READY;
 		return true;
 	}
-	if (!step) {
-		step = path;
-		timeStartStep = time;
-	} else if (step->Next && (( time - timeStartStep ) >= walk_speed)) {
-		//print("[New Step] : Orientation = %d", step->orient);
-		step = step->Next;
-		timeStartStep = timeStartStep + walk_speed;
-	}
-	SetOrientation (step->orient, true);
+
 	StanceID = IE_ANI_WALK;
 	if ((Type == ST_ACTOR) && (InternalFlags & IF_RUNNING)) {
 		StanceID = IE_ANI_RUN;
 	}
-	Pos.x = ( step->x * 16 ) + 8;
-	Pos.y = ( step->y * 12 ) + 6;
-	if (!step->Next) {
-		// we reached our destination, we are done
-		ClearPath();
-		NewOrientation = Orientation;
-		//since clearpath no longer sets currentaction to NULL
-		//we set it here
-		//no we don't, action is responsible for releasing itself
-		//ReleaseCurrentAction();
-		return true;
+
+	double dx, dxOrig;
+	double dy, dyOrig;
+	char ySign;
+	char xSign;
+	GetDeltas(dx, dy, ySign, xSign);
+	dxOrig = dx;
+	dyOrig = dy;
+	Point smPos = Point(Pos.x / 16, Pos.x / 12);
+	Point smStep = Point(step->x, step->y);
+
+	unsigned connectedWaypointsHeuristic = Distance(smPos, smStep);
+
+	unsigned timeDelta = time - timeStartStep;
+	if (timeDelta == 0) return true;
+
+	bool stepDoneWithWalkSpeed = timeDelta * connectedWaypointsHeuristic > walkAnimFrames;
+	if (stepDoneWithWalkSpeed) timeStartStep = time;
+
+	bool reachedStep = dx == 0 && dy == 0;
+	if (reachedStep) {
+		if (!step->Next) {
+			ClearPath();
+			NewOrientation = Orientation;
+			return true;
+		} else {
+			step = step->Next;
+			timeStartStep += walkAnimFrames;
+			return false;
+		}
 	}
-	if (( time - timeStartStep ) >= walk_speed) {
-		// we didn't finish all pending steps, yet
-		return false;
+
+	if (dx == 0) {
+		dy = maxRadius;
+	} else if (dy == 0) {
+		dx = maxRadius;
+	} else {
+		double squaredRadius = dx * dx + dy * dy;
+		double ratio = (maxRadius * maxRadius) / squaredRadius;
+		dx = sqrt(dx * dx * ratio);
+		dy = sqrt(dy * dy * ratio);
 	}
-	AdjustPositionTowards(Pos, time - timeStartStep, walk_speed, step->x, step->y, step->Next->x, step->Next->y);
-	return true;
+	dx = std::min(dx * stepLength / walkAnimFrames, dxOrig);
+	dy = std::min(dy * stepLength / walkAnimFrames, dyOrig);
+
+	//if (isInParty) Log(DEBUG, logString, "Moving by (%.2lf %.2lf)", dx, dy);
+
+	Pos.x += std::ceil(dx) * xSign;
+	Pos.y += std::ceil(dy) * ySign;
+
+	SetOrientation(step->orient, false);
+	return stepDoneWithWalkSpeed;
+}
+
+void Movable::GetDeltas(double &dx, double &dy, char &ySign, char &xSign) const {
+	dx = step->x * 16 + 16 - Pos.x;
+	dy = step->y * 12 + 12 - Pos.y;
```

Comment:
Interesting, the same +1 cell as in the pathfinder. What did the two additions solve?